rest_framework_json_api/filters/filter.py

```diff
@@ -0,0 +1,135 @@
+            `?filter[qty]=123`
+        - Apply other [field lookup](https://docs.djangoproject.com/en/stable/ref/models/querysets/#field-lookups)  # noqa: E501
+          operators:
+            `?filter[name.icontains]=bar` or `?filter[name.isnull]=true...`
+        - Membership in a list of values:
+            `?filter[name.in]=abc,123,zzz (name in ['abc','123','zzz'])`
+        - Filters can be combined for intersection (AND):
+            `?filter[qty]=123&filter[name.in]=abc,123,zzz&filter[...]`
+        - A related resource path can be used:
+        `?filter[inventory.item.partNum]=123456 (where `inventory.item` is the relationship path)`
+
+        If you are also using rest_framework.filters.SearchFilter you'll want to customize
+        the name of the query parameter for searching to make sure it doesn't conflict
+        with a field name defined in the filterset.
+        The recommended value is: `search_param="filter[search]"` but just make sure it's
+        `filter[<something>]` to comply with the jsonapi spec requirement to use the filter
+        keyword. The default is "search" unless overriden but it's used here just to make sure
+        we don't complain about it being an invalid filter.
+        """
+        # TODO: find a better way to deal with search_param.
+        search_param = api_settings.SEARCH_PARAM
+
+        # Make this regex check for 'filter' as well as 'filter[...]'
+        # Leave other incorrect usages of 'filter' to JSONAPIQueryValidationFilter.
+        # See http://jsonapi.org/format/#document-member-names for allowed characters
+        # and http://jsonapi.org/format/#document-member-names-reserved-characters for reserved
+        # characters (for use in paths, lists or as delimiters).
+        # regex `\w` matches [a-zA-Z0-9_].
+        # TODO: U+0080 and above allowed but not recommended. Leave them out for now. Fix later?
+        #       Also, ' ' (space) is allowed within a member name but not recommended.
+        filter_regex = re.compile(r'^filter(?P<ldelim>\[?)(?P<assoc>[\w\.\-]*)(?P<rdelim>\]?$)')
+
+        def validate_filter(self, keys, filterset_class):
+            for k in keys:
+                if ((not filterset_class) or (k not in filterset_class.base_filters)):
+                    raise ValidationError("invalid filter[{}]".format(k))
+
+        def get_filterset(self, request, queryset, view):
+            """
+            Sometimes there's no filterset_class defined yet the client still
+            requests a filter. Make sure they see an error too. This means
+            we have to get_filterset_kwargs() even if there's no filterset_class.
+
+            TODO: .base_filters vs. .filters attr (not always present)
+            """
+            filterset_class = self.get_filterset_class(view, queryset)
+            kwargs = self.get_filterset_kwargs(request, queryset, view)
+            self.validate_filter(self.filter_keys, filterset_class)
```

Comment:
we should keep this filter backend as a singleton and not introduce state.

What about `get_filterset_kwargs` returns filters as well? In `get_filterset` we can then do `kwargs.pop('filters')` so `filter_class(**kwargs) will still work?

Reply:
Oh you mean instead of carrying around `self.filter_keys`? Yeah, cool.